File: colleen/src/test/kotlin/io/github/cymoo/colleen/ScannerTest.java
package io.github.cymoo.colleen;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Method;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Test suite for ControllerScanner functionality
 */
@DisplayName("ControllerScanner Tests")
class ControllerScannerJavaTest {

    // ========== Test Controllers ==========

    @Controller("/api")
    static class BasicController {
        @Get("/users")
        public void getUsers(Context ctx) {
        }

        @Post("/users")
        public void createUser(Context ctx) {
        }

        @Put("/users/{id}")
        public void updateUser(Context ctx) {
        }

        @Delete("/users/{id}")
        public void deleteUser(Context ctx) {
        }
    }

    @Controller
    static class DefaultPathController {
        @Get("/home")
        public void home(Context ctx) {
        }
    }

    static class MultipleRoutesController {
        @Get("/")
        @Get("/home")
        @Get("/index")
        public void homePage(Context ctx) {
        }

        @Get("/user/{id}")
        @Put("/user/{id}")
        @Delete("/user/{id}")
        public void handleUser(Context ctx) {
        }
    }

    static class MiddlewareController {
        @Use
        public void authMiddleware(Context ctx, Next next) {
        }

        @Use
        public void loggingMiddleware(Context ctx, Next next) {
        }

        @Get("/protected")
        public void protectedRoute(Context ctx) {
        }
    }

    static class PathNormalizationController {
        @Get("")
        public void emptyPath(Context ctx) {
        }

        @Get("no-slash")
        public void noLeadingSlash(Context ctx) {
        }

        @Get("/trailing/")
        public void trailingSlash(Context ctx) {
        }

        @Get("/")
        public void rootPath(Context ctx) {
        }
    }

    // Invalid controllers for testing error cases
    static class InvalidMiddlewareController1 {
        @Use
        public void wrongParamCount(Context ctx) {
        } // Only 1 parameter
    }

    static class InvalidMiddlewareController2 {
        @Use
        public void wrongParamTypes(String str, Integer num) {
        } // Wrong types
    }

    static class InvalidMiddlewareController3 {
        @Use
        public String wrongReturnType(Context ctx, Next next) { // Returns String
            return "invalid";
        }
    }

    // ========== Basic Scanning Tests ==========

    @Nested
    @DisplayName("Basic Controller Scanning")
    class BasicScanningTests {

        @Test
        @DisplayName("Should scan basic controller with correct base path")
        void testBasicControllerBasePath() {
            BasicController controller = new BasicController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            assertEquals("/api", info.getBasePath());
            assertSame(controller, info.getObj());
        }

        @Test
        @DisplayName("Should discover all HTTP method routes")
        void testAllHttpMethods() {
            BasicController controller = new BasicController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            List<RouteInfo> routes = info.getRoutes();
            assertEquals(4, routes.size());

            var gets = routes.stream().filter(route -> route.getMethod().equals("GET") && route.getPath().equals("/users")).toList();
            assertEquals(1, gets.size());

            var posts = routes.stream().filter(route -> route.getMethod().equals("POST") && route.getPath().equals("/users")).toList();
            assertEquals(1, posts.size());

            var puts = routes.stream().filter(route -> route.getMethod().equals("PUT") && route.getPath().equals("/users/{id}")).toList();
            assertEquals(1, posts.size());

            var deletes = routes.stream().filter(route -> route.getMethod().equals("DELETE") && route.getPath().equals("/users/{id}")).toList();
            assertEquals(1, deletes.size());
        }

        @Test
        @DisplayName("Should use default path when @Controller has no value")
        void testDefaultPath() {
            DefaultPathController controller = new DefaultPathController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            assertEquals("/", info.getBasePath());
        }

        @Test
        @DisplayName("Should make handler methods accessible")
        void testMethodAccessibility() {
            BasicController controller = new BasicController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            for (RouteInfo route : info.getRoutes()) {
                assertTrue(route.getHandler().isAccessible(),
                        "Method " + route.getHandler().getName() + " should be accessible");
            }
        }
    }

    // ========== Multiple Routes Tests ==========

    @Nested
    @DisplayName("Multiple Routes Per Method")
    class MultipleRoutesTests {

        @Test
        @DisplayName("Should handle multiple @Get annotations on same method")
        void testMultipleGetAnnotations() {
            MultipleRoutesController controller = new MultipleRoutesController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            List<RouteInfo> homeRoutes = info.getRoutes().stream()
                    .filter(r -> r.getHandler().getName().equals("homePage"))
                    .toList();

            assertEquals(3, homeRoutes.size());

            List<String> paths = homeRoutes.stream()
                    .map(RouteInfo::getPath)
                    .sorted()
                    .collect(Collectors.toList());

            assertEquals(List.of("/", "/home", "/index"), paths);

            // All should be GET
            assertTrue(homeRoutes.stream().allMatch(r -> r.getMethod().equals("GET")));
        }

        @Test
        @DisplayName("Should handle multiple different HTTP methods on same method")
        void testMultipleDifferentHttpMethods() {
            MultipleRoutesController controller = new MultipleRoutesController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            List<RouteInfo> userRoutes = info.getRoutes().stream()
                    .filter(r -> r.getHandler().getName().equals("handleUser"))
                    .toList();

            assertEquals(3, userRoutes.size());

            Map<String, String> methodMap = userRoutes.stream()
                    .collect(Collectors.toMap(RouteInfo::getMethod, RouteInfo::getPath));

            assertEquals("/user/{id}", methodMap.get("GET"));
            assertEquals("/user/{id}", methodMap.get("PUT"));
            assertEquals("/user/{id}", methodMap.get("DELETE"));
        }

        @Test
        @DisplayName("Should refer to same Method instance for routes from same handler")
        void testSameMethodInstance() {
            MultipleRoutesController controller = new MultipleRoutesController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            List<Method> homeMethods = info.getRoutes().stream()
                    .map(RouteInfo::getHandler)
                    .filter(handler -> handler.getName().equals("homePage"))
                    .distinct()
                    .toList();

            assertEquals(1, homeMethods.size(), "All routes should reference the same Method object");
        }
    }

    // ========== Middleware Tests ==========

    @Nested
    @DisplayName("Middleware Discovery")
    class MiddlewareTests {

        @Test
        @DisplayName("Should discover all @Use middleware methods")
        void testMiddlewareDiscovery() {
            MiddlewareController controller = new MiddlewareController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            List<Method> middlewares = info.getMiddlewares();
            assertEquals(2, middlewares.size());

            List<String> names = middlewares.stream()
                    .map(Method::getName)
                    .sorted()
                    .collect(Collectors.toList());

            assertEquals(List.of("authMiddleware", "loggingMiddleware"), names);
        }

        @Test
        @DisplayName("Should make middleware methods accessible")
        void testMiddlewareAccessibility() {
            MiddlewareController controller = new MiddlewareController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            for (Method middleware : info.getMiddlewares()) {
                assertTrue(middleware.isAccessible(),
                        "Middleware " + middleware.getName() + " should be accessible");
            }
        }

        @Test
        @DisplayName("Should not confuse middleware with routes")
        void testMiddlewareSeparation() {
            MiddlewareController controller = new MiddlewareController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            assertEquals(2, info.getMiddlewares().size());
            assertEquals(1, info.getRoutes().size());
            assertEquals("/protected", info.getRoutes().getFirst().getPath());
        }
    }

    // ========== Error Handling Tests ==========

    @Nested
    @DisplayName("Error Handling and Validation")
    class ErrorHandlingTests {

        @Test
        @DisplayName("Should throw when middleware has wrong parameter count")
        void testMiddlewareWrongParamCount() {
            InvalidMiddlewareController1 controller = new InvalidMiddlewareController1();

            IllegalArgumentException exception = assertThrows(
                    IllegalArgumentException.class,
                    () -> ControllerScanner.INSTANCE.scan(controller)
            );

            assertTrue(exception.getMessage().contains("must have exactly 2 parameters"));
            assertTrue(exception.getMessage().contains("wrongParamCount"));
        }

        @Test
        @DisplayName("Should throw when middleware has wrong parameter types")
        void testMiddlewareWrongParamTypes() {
            InvalidMiddlewareController2 controller = new InvalidMiddlewareController2();

            IllegalArgumentException exception = assertThrows(
                    IllegalArgumentException.class,
                    () -> ControllerScanner.INSTANCE.scan(controller)
            );

            assertTrue(exception.getMessage().contains("must be Context") ||
                    exception.getMessage().contains("must be Next"));
        }

        @Test
        @DisplayName("Should throw when middleware has wrong return type")
        void testMiddlewareWrongReturnType() {
            InvalidMiddlewareController3 controller = new InvalidMiddlewareController3();

            IllegalArgumentException exception = assertThrows(
                    IllegalArgumentException.class,
                    () -> ControllerScanner.INSTANCE.scan(controller)
            );

            assertTrue(exception.getMessage().contains("must be void/Unit"));
            assertTrue(exception.getMessage().contains("wrongReturnType"));
        }

        @Test
        @DisplayName("Should handle controller with no routes or middlewares")
        void testEmptyController() {
            @Controller("/empty")
            class EmptyController {
            }

            EmptyController controller = new EmptyController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            assertEquals("/empty", info.getBasePath());
            assertTrue(info.getRoutes().isEmpty());
            assertTrue(info.getMiddlewares().isEmpty());
        }
    }

    // ========== Integration Tests ==========

    @Nested
    @DisplayName("Integration Scenarios")
    class IntegrationTests {

        @Test
        @DisplayName("Should handle complex controller with everything")
        void testComplexController() {
            @Controller("/api/v1")
            class ComplexController {
                @Use
                public void auth(Context ctx, Next next) {
                }

                @Use
                public void logging(Context ctx, Next next) {
                }

                @Get
                @Get("/index")
                public void home(Context ctx) {
                }

                @Get("/users")
                public void listUsers(Context ctx) {
                }

                @Post("/users")
                public void createUser(Context ctx) {
                }

                @Get("/users/{id}")
                @Put("/users/{id}")
                @Delete("/users/{id}")
                public void handleUser(Context ctx) {
                }
            }

            ComplexController controller = new ComplexController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            assertEquals("/api/v1", info.getBasePath());
            assertEquals(2, info.getMiddlewares().size());
            assertEquals(7, info.getRoutes().size()); // 2 + 1 + 1 + 3

            // Verify route distribution
            Map<String, Long> methodCounts = info.getRoutes().stream()
                    .collect(Collectors.groupingBy(RouteInfo::getMethod, Collectors.counting()));

            assertEquals(4L, methodCounts.get("GET"));
            assertEquals(1L, methodCounts.get("PUT"));
            assertEquals(1L, methodCounts.get("POST"));
            assertEquals(1L, methodCounts.get("DELETE"));
        }

        @Test
        @DisplayName("Should preserve method metadata")
        void testMethodMetadataPreservation() {
            BasicController controller = new BasicController();
            ControllerInfo info = ControllerScanner.INSTANCE.scan(controller);

            RouteInfo getUsersRoute = info.getRoutes().stream()
                    .filter(r -> r.getPath().equals("/users") && r.getMethod().equals("GET"))
                    .findFirst()
                    .orElseThrow();

            Method handler = getUsersRoute.getHandler();
            assertEquals("getUsers", handler.getName());
            assertEquals(1, handler.getParameterCount());
            assertEquals(Context.class, handler.getParameterTypes()[0]);
        }
    }
}

File: colleen/src/test/kotlin/io/github/cymoo/colleen/ScannerTest.kt
package io.github.cymoo.colleen

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

@Controller("/api")
class TestController {
    @Use
    fun authMiddleware(ctx: Context, next: Next) {
        // middleware logic
    }

    @Get("/users")
    fun getUsers() = "users"

    @Post("/users")
    fun createUser() = "create"

    @Put("/users/:id")
    fun updateUser() = "update"

    @Delete("/users/:id")
    fun deleteUser() = "delete"

    fun helperMethod() = "helper"
}

@Controller
class DefaultPathController {
    @Get("/home")
    fun home() = "home"
}

class NoControllerAnnotation {
    @Get("/test")
    fun test() = "test"
}

@Controller("/admin")
class MultipleMiddlewaresController {
    @Use
    fun firstMiddleware(ctx: Context, next: Next) {
    }

    @Use
    fun secondMiddleware(ctx: Context, next: Next) {
    }

    @Get("/dashboard")
    fun dashboard() = "dashboard"
}

@Controller("/missing-next")
class InvalidMiddlewareController1 {
    @Use
    fun badMiddleware(ctx: Context) {
    }
}

@Controller("/invalid-parameter-types")
class InvalidMiddlewareController2 {
    @Use
    fun badMiddleware(str: String, num: Int) {
    }
}

@Controller("/invalid-return-type")
class InvalidMiddlewareController3 {
    @Use
    fun badMiddleware(ctx: Context, next: Next): String {
        return "invalid"
    }
}

class ControllerScannerTest {

    @Test
    fun `test scan basic controller with routes`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        assertEquals("/api", info.basePath)
        assertEquals(1, info.middlewares.size)
        assertEquals(4, info.routes.size)
        assertEquals(controller, info.obj)
    }

    @Test
    fun `test scan finds correct base path`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        assertEquals("/api", info.basePath)
    }

    @Test
    fun `test scan with default base path`() {
        val controller = DefaultPathController()
        val info = ControllerScanner.scan(controller)

        assertEquals("/", info.basePath)
    }

    @Test
    fun `test scan without controller annotation`() {
        val controller = NoControllerAnnotation()
        val info = ControllerScanner.scan(controller)

        assertEquals("/", info.basePath)
    }

    @Test
    fun `test finds all HTTP method routes`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        val routes = info.routes
        assertEquals(4, routes.size)

        val getRoute = routes.find { it.method == "GET" }
        assertNotNull(getRoute)
        assertEquals("/users", getRoute?.path)
        assertEquals("getUsers", getRoute?.handler?.name)

        val postRoute = routes.find { it.method == "POST" }
        assertNotNull(postRoute)
        assertEquals("/users", postRoute?.path)
        assertEquals("createUser", postRoute?.handler?.name)

        val putRoute = routes.find { it.method == "PUT" }
        assertNotNull(putRoute)
        assertEquals("/users/:id", putRoute?.path)
        assertEquals("updateUser", putRoute?.handler?.name)

        val deleteRoute = routes.find { it.method == "DELETE" }
        assertNotNull(deleteRoute)
        assertEquals("/users/:id", deleteRoute?.path)
        assertEquals("deleteUser", deleteRoute?.handler?.name)
    }


    @Test
    fun `test finds all HTTP method routes of a method`() {
        @Controller("/base")
        class MyController {
            @Get
            @Get("/home")
            fun index() = "index"
        }

        val info = ControllerScanner.scan(MyController())

        assertEquals(2, info.routes.size)
    }

    @Test
    fun `test finds middlewares`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        assertEquals(1, info.middlewares.size)
        assertEquals("authMiddleware", info.middlewares[0].name)
    }

    @Test
    fun `test finds multiple middlewares`() {
        val controller = MultipleMiddlewaresController()
        val info = ControllerScanner.scan(controller)

        assertEquals(2, info.middlewares.size)
        assertTrue(info.middlewares.any { it.name == "firstMiddleware" })
        assertTrue(info.middlewares.any { it.name == "secondMiddleware" })
    }

    @Test
    fun `test validates middleware signature - wrong parameter count`() {
        val controller = InvalidMiddlewareController1()

        val exception = assertThrows<IllegalArgumentException> {
            ControllerScanner.scan(controller)
        }

        assertTrue(exception.message!!.contains("must have exactly 2 parameters"))
    }

    @Test
    fun `test validates middleware signature - wrong parameter types`() {
        val controller = InvalidMiddlewareController2()

        val exception = assertThrows<IllegalArgumentException> {
            ControllerScanner.scan(controller)
        }

        assertTrue(exception.message!!.contains("First parameter must be Context"))
    }

    @Test
    fun `test validates middleware signature - wrong return type`() {
        val controller = InvalidMiddlewareController3()

        val exception = assertThrows<IllegalArgumentException> {
            ControllerScanner.scan(controller)
        }

        assertTrue(exception.message!!.contains("Return type must be void/Unit"))
    }

    @Test
    fun `test route handler is accessible`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        info.routes.forEach { route ->
            assertTrue(route.handler.trySetAccessible())
        }
    }

    @Test
    fun `test middleware is accessible`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        info.middlewares.forEach { middleware ->
            assertTrue(middleware.trySetAccessible())
        }
    }

    @Test
    fun `test ignores non-annotated methods`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        // helperMethod不应该被识别为路由
        val helperRoute = info.routes.find { it.handler.name == "helperMethod" }
        assertEquals(null, helperRoute)
    }

    @Test
    fun `test controller info contains correct object reference`() {
        val controller = TestController()
        val info = ControllerScanner.scan(controller)

        assertEquals(controller, info.obj)
        assertTrue(info.obj is TestController)
    }

    @Test
    fun `test empty routes when no HTTP annotations`() {
        @Controller("/empty")
        class EmptyController

        val controller = EmptyController()
        val info = ControllerScanner.scan(controller)

        assertEquals(0, info.routes.size)
        assertEquals(0, info.middlewares.size)
    }

    @Test
    fun `test route with empty path`() {
        @Controller("/base")
        class EmptyPathController {
            @Get
            fun index() = "index"
        }

        val controller = EmptyPathController()
        val info = ControllerScanner.scan(controller)

        assertEquals(1, info.routes.size)
        assertEquals("/", info.routes[0].path)
    }
}